helper/tlsutil/generate.go

```diff
@@ -81,57 +89,122 @@ type CertOpts struct {
 	ExtKeyUsage []x509.ExtKeyUsage
 }
 
+// Check if any of CAOpts the variables are populated
+func AreCARequiredFieldsEmpty(caOpts CAOpts) bool {
+	return caOpts.Days == 0 &&
+		caOpts.Country == "" &&
+		caOpts.PostalCode == "" &&
+		caOpts.Province == "" &&
+		caOpts.Locality == "" &&
+		caOpts.StreetAddress == "" &&
+		caOpts.Organization == "" &&
+		caOpts.OrganizationalUnit == "" &&
+		caOpts.Name == ""
+}
+
 // GenerateCA generates a new CA for agent TLS (not to be confused with Connect TLS)
 func GenerateCA(opts CAOpts) (string, string, error) {
-	signer := opts.Signer
-	var pk string
-	if signer == nil {
-		var err error
-		signer, pk, err = GeneratePrivateKey()
+	var (
+		id     []byte
+		pk     string
+		err    error
+		signer = opts.Signer
+		sn     = opts.Serial
+	)
+
+	// Check if opts required fields are empty
+	isEmpty := AreCARequiredFieldsEmpty(opts)
+	if isEmpty {
+
+		if signer == nil {
+			var err error
+			signer, pk, err = GeneratePrivateKey()
+			if err != nil {
+				return "", "", err
+			}
+		}
+
+		certKeyId, err := keyID(signer.Public())
 		if err != nil {
 			return "", "", err
 		}
-	}
-
-	id, err := keyID(signer.Public())
-	if err != nil {
-		return "", "", err
-	}
+		id = certKeyId
+
+		if sn == nil {
+			var err error
+			sn, err = GenerateSerialNumber()
+			if err != nil {
+				return "", "", err
+			}
+		}
+		opts.Name = fmt.Sprintf("Nomad Agent CA %d", sn)
+		opts.Days = 1825
+		opts.Country = "US"
+		opts.PostalCode = "94105"
+		opts.Province = "CA"
+		opts.Locality = "San Francisco"
+		opts.StreetAddress = "101 Second Street"
+		opts.Organization = "HashiCorp Inc."
+		opts.OrganizationalUnit = "Nomad"
+
+	} else {
+		if signer == nil {
+			var err error
+			signer, pk, err = GeneratePrivateKey()
+			if err != nil {
+				return "", "", err
+			}
+		}
 
-	sn := opts.Serial
-	if sn == nil {
-		var err error
-		sn, err = GenerateSerialNumber()
+		certKeyId, err := keyID(signer.Public())
 		if err != nil {
 			return "", "", err
 		}
-	}
-	name := opts.Name
-	if name == "" {
-		name = fmt.Sprintf("Nomad Agent CA %d", sn)
-	}
+		id = certKeyId
+
+		if sn == nil {
+			var err error
+			sn, err = GenerateSerialNumber()
+			if err != nil {
+				return "", "", err
+			}
+		}
 
-	days := opts.Days
-	if opts.Days == 0 {
-		days = 365
+		if opts.Name == "" {
+			return "", "", errors.New("please provide the -common-name flag when customizing the CA")
+		} else {
+			opts.Name = fmt.Sprintf("%s %d", opts.Name, sn)
+		}
+		if opts.Country == "" {
+			return "", "", errors.New("please provide the -country flag when customizing the CA")
+		}
+
+		if opts.Organization == "" {
+			return "", "", errors.New("please provide the -organization flag when customizing the CA")
+		}
+
+		if opts.OrganizationalUnit == "" {
+			return "", "", errors.New("please provide the -organizational-unit flag when customizing the CA")
```

Comment:
The errors are specific to the CLI implementation; I wonder if we want to make these more generic, so future internal callers do not emit errors related to CLI flags when they do not implement them?